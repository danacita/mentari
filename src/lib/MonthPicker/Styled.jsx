import styled, { css } from 'styled-components';

import Color from '../Styles/bases/Color';
import { Transition } from '../Styles/utils';

const MonthWrapper = styled.div`
  .ryms-container {
    width: 300px;
    z-index: 100;
    background: #fff;
    border: 2px solid #eee;
    border-bottom-right-radius: 5px;
    border-bottom-left-radius: 5px;
    position: absolute;
  }
  .ryms-clickout {
    width: 100vw;
    height: 100vh;
    position: fixed;
    top: 0;
    left: 0;
    z-index: -1;
    background: transparent;
    display: none;
  }
  /*  SECTIONS  */
  .ryms-section_mp {
    clear: both;
    padding: 0px;
    margin: 0px;
  }

  /*  COLUMN SETUP  */
  .ryms-col_mp {
    display: block;
    float: left;
    text-align: center;
  }
  .ryms-col_mp:first-child {
    margin-left: 0;
  }

  /*  GROUPING  */
  .ryms-group_mp:before,
  .group:after {
    content: '';
    display: table;
  }
  .ryms-group_mp:after {
    clear: both;
  }
  .ryms-group_mp {
    zoom: 1; /* For IE 6/7 */
  }

  /*  GRID OF THREE  */
  .ryms-span_1_of_3_mp {
    width: 33.33%;
  }

  .ryms-col_mp {
    font-size: 16px;
    padding-bottom: 12px;
    padding-top: 12px;
  }

  .ryms-col_mp:hover {
    color: ${Color.black};
    background-color: ${Color.grey239};
    cursor: pointer;
  }
  .ryms-selected_date_mp {
    font-size: 16px;
    color: ${Color.darkBlue};
    font-weight: bold;
    padding-top: 13px;
  }
  .ryms-selected_cell {
    background-color: ${Color.blue};
    font-weight: 600;
    font-style: normal;
    color: ${Color.white};
    border-radius: 4px;
    &:hover {
      background-color: ${Color.blue};
      color: ${Color.white};
    }
  }
  .ryms-arrows_mp {
    font-weight: bold;
    font-size: 18px;
  }

  position: relative;
  font-size: 16px;
  input:-webkit-autofill {
    box-shadow: 0 0 0px 1000px white inset;
    -webkit-box-shadow: 0 0 0px 1000px white inset;
  }
  input[type='text'],
  input[type='number'],
  input[type='email'],
  input[type='password'] {
    position: relative;
    border: 1px solid ${Color.grey191};
    line-height: 1.6;
    display: block;
    min-height: 49px;
    max-height: 49px;
    width: 100%;
    padding: 8px 16px;
    border-radius: 4px;
    color: white;
    ${Transition('all ease-in-out 0.2s')} outline: none;
    &::placeholder {
      color: white;
    }
    &:focus,
    &:active {
      border-color: ${Color.blue};
      padding-top: 14px;
      + label {
        top: -10px;
        color: ${Color.blue};
        font-size: 12px;
        font-weight: 600;
      }
      &::placeholder {
        color: ${Color.grey191};
      }
    }
    + label {
      position: absolute;
      z-index: 20;
      left: 8px;
      background-color: ${Color.white};
      padding: 0 8px;
      top: 14px;
      ${Transition('all ease-in-out 0.2s')};
      cursor: text;
      color: ${Color.grey191};
      height: 24px;
      display: flex;
      align-items: center;
    }
    &:disabled {
      background-color: ${Color.white};
      cursor: not-allowed;
      color: ${Color.grey216};
      border-color: ${Color.grey239};
      &::placeholder {
        color: ${Color.grey239};
        opacity: 0;
      }
      + label {
        background-color: ${Color.white};
        color: ${Color.grey216};
        cursor: not-allowed;
      }
      &:active {
        padding-top: 14px;
        border-color: ${Color.grey191};
        + label {
          top: 12px;
          color: ${Color.grey191};
          font-size: 14px;
          font-weight: 400;
        }
        &::placeholder {
          color: ${Color.grey191};
        }
      }
    }
  }
  ${props =>
    props.valLength > 0
      ? css`
          input[type='text'],
          input[type='email'],
          input[type='number'],
          input[type='password'] {
            padding-top: 14px;
            color: black;
            + label {
              top: -10px;
              color: ${Color.grey191};
              font-size: 12px;
              font-weight: 600;
            }
          }
        `
      : null};
  ${props =>
    props.touched && props.error
      ? css`
          input[type='text'],
          input[type='email'],
          input[type='number'],
          input[type='password'] {
            border: 1px solid ${Color.red};
            &:focus,
            &:hover,
            &:active {
              border-color: ${Color.red};
              + label {
                color: ${Color.red};
              }
            }
            + label {
              color: ${Color.red};
            }
          }
        `
      : null};
  ${props =>
    props.hasPrefix
      ? css`
          input[type='text'],
          input[type='email'],
          input[type='number'],
          input[type='password'] {
            padding-left: 44px;
          }
        `
      : null};
`;

const MonthPrefix = styled.div`
  position: absolute;
  z-index: 2;
  top: 14px;
  left: 14px;
`;

export { MonthWrapper, MonthPrefix };
