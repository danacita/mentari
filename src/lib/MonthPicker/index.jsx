/* eslint-disable no-unused-expressions */
/* eslint-disable no-lone-blocks */
/* eslint-disable no-return-assign */
/* eslint-disable jsx-a11y/label-has-for */
import moment from 'moment';
import PropTypes from 'prop-types';
import React, { Component, Fragment } from 'react';
import YearMonthSelector from 'react-year-month-selector';
import { MonthWrapper } from './Styled';
import { Margin } from '../Styles/utils';
import Color from '../Styles/bases/Color';

const propTypes = {
  id: PropTypes.string,
  placeholder: PropTypes.string,
  defaultPickerValue: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string,
  disabled: PropTypes.bool,
  withRedux: PropTypes.bool,
  // input prop is passed by redux-form to keep field in sync with state
  input: PropTypes.shape({
    value: PropTypes.string,
    onChange: PropTypes.func
  }),
  meta: PropTypes.shape({
    error: PropTypes.string,
    touched: PropTypes.bool
  })
};

const defaultProps = {
  id: null,
  placeholder: null,
  label: null,
  name: null,
  disabled: false,
  defaultPickerValue: null,
  withRedux: true,
  input: {},
  meta: {}
};

class MonthPicker extends Component {
  constructor(props) {
    super(props);
    const { defaultPickerValue } = props;
    this.state = {
      pickerOpen: false,
      valueOfState: '',
      month:
        defaultPickerValue !== null
          ? Number(moment(defaultPickerValue).format('MM')) - 1
          : new Date().getMonth(),
      year:
        defaultPickerValue !== null
          ? Number(moment(defaultPickerValue).format('YYYY'))
          : new Date().getFullYear()
    };
    this.handleOpen = this.handleOpen.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  handleOpen() {
    this.setState({
      pickerOpen: true
    });
  }

  handleClose() {
    this.setState({ pickerOpen: false });
  }

  handleChange(year, month) {
    const { input, withRedux } = this.props;
    {
      withRedux &&
        input.onChange(
          year !== ''
            ? moment()
                .year(year)
                .month(month)
                .format('YYYY-MM')
            : ''
        );
    }
    this.setState({
      month,
      year,
      valueOfState: `${year}-${month + 1}`
    });
  }

  handleClickOutside(e) {
    if (this.node !== null && !this.node.contains(e.target)) {
      this.setState({
        pickerOpen: false
      });
    }
  }

  render() {
    const {
      id,
      name,
      placeholder,
      disabled,
      label,
      withRedux,
      input,
      meta: { error, touched }
    } = this.props;
    const { pickerOpen, month, year, valueOfState } = this.state;
    return (
      <Fragment>
        {withRedux ? (
          <MonthWrapper
            valLength={input.value && input.value.length}
            touched={touched}
            error={error}
          >
            <input
              id={id}
              data-testid={id}
              name={name}
              type="text"
              placeholder={placeholder}
              onClick={this.handleOpen}
              onClose={this.handleClose}
              value={input.value !== 'Invalid date' ? input.value : ''}
              autoComplete="off"
              onChange={input.onChange}
              disabled={disabled}
              readOnly
            />
            <label htmlFor={id}>{label}</label>
            <div ref={node => (this.node = node)}>
              <YearMonthSelector
                year={year}
                month={month}
                open={pickerOpen}
                onClose={this.handleClose}
                onChange={this.handleChange}
              />
            </div>
            {touched && error && (
              <Margin top={8}>
                <span style={{ color: Color.red }}>{error}</span>
              </Margin>
            )}
          </MonthWrapper>
        ) : (
          <MonthWrapper valLength={valueOfState && valueOfState.length}>
            <input
              id={id}
              data-testid={id}
              name={name}
              type="text"
              placeholder={placeholder}
              onClick={this.handleOpen}
              onClose={this.handleClose}
              value={valueOfState}
              autoComplete="off"
              onChange={this.handleChangevalueOfState}
              disabled={disabled}
              readOnly
            />
            <label htmlFor={id}>{label}</label>
            <div ref={node => (this.node = node)}>
              <YearMonthSelector
                year={year}
                month={month}
                open={pickerOpen}
                onClose={this.handleClose}
                onChange={this.handleChange}
              />
            </div>
          </MonthWrapper>
        )}
      </Fragment>
    );
  }
}

MonthPicker.propTypes = propTypes;
MonthPicker.defaultProps = defaultProps;

export default MonthPicker;
