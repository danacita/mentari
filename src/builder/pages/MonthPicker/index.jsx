import React, { Component } from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { atomOneDarkReasonable } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import { ContentWithRightBar } from '../Styled';
import RightbarShared from '../../shared/Rightbar';
import DemoShared from '../../shared/Demo';
import PropsShared from '../../shared/Props';
import CodeShared from '../../shared/Code';
import { Margin, Checkbox, MonthPicker, Row, Col, TextField } from '../../../lib';

class MonthPickerPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      handleMonthPickerId: 'Id',
      handleMonthPickerName: 'Name',
      handleMonthPickerLabel: 'Label',
      handleMonthPickerPlaceholder: 'Placeholder',
      handleMonthPickerDisable: false
    };
  }

  render() {
    const {
      handleMonthPickerLabel,
      handleMonthPickerId,
      handleMonthPickerName,
      handleMonthPickerPlaceholder,
      handleMonthPickerDisable
    } = this.state;
    const codeString = `import { MonthPicker } from '@danacita/mentari';\n<MonthPicker${
      handleMonthPickerId ? ` id="${handleMonthPickerId}"` : ''
    }${handleMonthPickerName ? ` name="${handleMonthPickerName}"` : ''}${
      handleMonthPickerLabel ? ` label="${handleMonthPickerLabel}"` : ''
    }${handleMonthPickerPlaceholder ? ` placeholder="${handleMonthPickerPlaceholder}"` : ''}${
      handleMonthPickerDisable ? ` disabled` : ''
    } />`;
    return (
      <ContentWithRightBar>
        <h1>Monthpicker</h1>
        <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
        <Margin top={24}>
          <PropsShared>
            <li>
              <span>id</span>
              <span>Apply a id to the control element</span>
            </li>
            <li>
              <span>name</span>
              <span>Defines name input Monthpicker</span>
            </li>
            <li>
              <span>label</span>
              <span>Defines label name Textfield</span>
            </li>
            <li>
              <span>placeholder</span>
              <span>Apply a short hint that describes the expected value when input focus</span>
            </li>
            <li>
              <span>disabled</span>
              <span>Apply disabled state of Textfield for unusable and unclickable</span>
            </li>
          </PropsShared>
        </Margin>
        <DemoShared>
          <h4>Demo</h4>
          <p>You can change the props via the component builder on your right</p>
          <Margin top={16}>
            <Row>
              <Col lg={4}>
                <MonthPicker
                  id={handleMonthPickerId}
                  label={handleMonthPickerLabel}
                  placeholder={handleMonthPickerPlaceholder}
                  withRedux={false}
                  disabled={handleMonthPickerDisable}
                />
              </Col>
            </Row>
          </Margin>
        </DemoShared>
        <h4>Code</h4>
        <Margin top={16}>
          <CodeShared copyCode={codeString}>
            <SyntaxHighlighter language="jsx" style={atomOneDarkReasonable} showLineNumbers>
              {codeString}
            </SyntaxHighlighter>
          </CodeShared>
        </Margin>
        <RightbarShared title="Monthpicker">
          <Margin bottom={24}>
            <TextField
              id="handleMonthPickerId"
              label="Monthpicker id"
              autoFocus
              withRedux={false}
              value={handleMonthPickerId}
              onChange={e => this.setState({ handleMonthPickerId: e.target.value })}
            />
          </Margin>
          <Margin bottom={24}>
            <TextField
              id="handleMonthPickerName"
              label="Monthpicker name"
              autoFocus
              withRedux={false}
              value={handleMonthPickerName}
              onChange={e => this.setState({ handleMonthPickerName: e.target.value })}
            />
          </Margin>
          <Margin bottom={24}>
            <TextField
              id="handleMonthPickerLabel"
              label="Monthpicker label"
              withRedux={false}
              value={handleMonthPickerLabel}
              onChange={e => this.setState({ handleMonthPickerLabel: e.target.value })}
            />
          </Margin>
          <Margin bottom={24}>
            <TextField
              id="handleMonthPickerPlaceholder"
              label="Monthpicker placeholder"
              withRedux={false}
              value={handleMonthPickerPlaceholder}
              onChange={e => this.setState({ handleMonthPickerPlaceholder: e.target.value })}
            />
          </Margin>
          <Margin bottom={16}>
            <h4>State of Monthpicker</h4>
          </Margin>
          <Margin bottom={24}>
            <Checkbox
              id="handleMonthPickerDisable"
              name="handleMonthPickerDisable"
              label="Disable"
              checked={handleMonthPickerDisable}
              withRedux={false}
              onChange={e => this.setState({ handleMonthPickerDisable: e.target.checked })}
            />
          </Margin>
        </RightbarShared>
      </ContentWithRightBar>
    );
  }
}

export default MonthPickerPage;
