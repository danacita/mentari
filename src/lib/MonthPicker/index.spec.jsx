import React from 'react';
import renderer from 'react-test-renderer';

import MonthPicker from '.';

describe('MonthPicker', () => {
  it('renders correctly', () => {
    const input = {
      value: 'Value',
      onChange: () => {}
    };

    const meta = {
      error: '',
      touched: false
    };

    const tree = renderer.create(<MonthPicker input={input} meta={meta} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
