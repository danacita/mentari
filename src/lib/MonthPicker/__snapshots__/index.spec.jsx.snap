// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthPicker renders correctly 1`] = `
<div
  className="sc-htpNat ihUAIK"
>
  <input
    autoComplete="off"
    data-testid={null}
    disabled={false}
    id={null}
    name={null}
    onChange={[Function]}
    onClick={[Function]}
    onClose={[Function]}
    placeholder={null}
    readOnly={true}
    type="text"
    value="Value"
  />
  <label
    htmlFor={null}
  />
  <div>
    <div />
  </div>
</div>
`;
